In [12]:
from Utils.createURM import createURM
import pandas as pd
import numpy as np
import scipy.sparse as sp

In [13]:
dataset = pd.read_csv('../Input/interactions_and_impressions.csv')
dataset = dataset.drop(columns=['Impressions'])

datasetCOO = sp.coo_matrix((dataset["Data"].values, (dataset["UserID"].values, dataset["ItemID"].values)))
userIDS = dataset['UserID'].unique()
itemIDS = dataset['ItemID'].unique()

URM = np.zeros((len(userIDS), len(itemIDS)), dtype=int)
for x in range(len(datasetCOO.data)):
    if datasetCOO.data[x] == 0:
        URM[datasetCOO.row[x]][datasetCOO.col[x]] = int(1)

URM

/var/folders/tt/9v6608x550n51759gpcrffsc0000gn/T/ipykernel_15183/2492280762.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  dataset = pd.read_csv('../Input/interactions_and_impressions.csv')


array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [14]:
URM.shape

(41629, 24507)

In [15]:
lenght = pd.read_csv('../Input/data_ICM_length.csv')
lenght['item_id']

0            0
1            1
2            2
3            3
4            4
         ...  
23086    27963
23087    27964
23088    27965
23089    27966
23090    27967
Name: item_id, Length: 23091, dtype: int64

In [10]:
types = pd.read_csv('../Input/data_ICM_type.csv')
types

item_id  feature_id  data
0            0           1     1
1            1           3     1
2            2           4     1
3            3           1     1
4            4           3     1
...        ...         ...   ...
23086    27963           1     1
23087    27964           2     1
23088    27965           1     1
23089    27966           1     1
23090    27967           4     1

[23091 rows x 3 columns]

## Startegia
Una possibile strategia potrebbe essere quella di aumentare le colonne della URM fino all'item 27967 per poi trainare eventuali valori.

In [31]:
dataset = pd.read_csv('../Input/interactions_and_impressions.csv')
dataset = dataset.drop(columns=['Impressions'])

datasetCOO = sp.coo_matrix((dataset["Data"].values, (dataset["UserID"].values, dataset["ItemID"].values)))
userIDS = dataset['UserID'].unique()
itemIDS = dataset['ItemID'].unique()

newURM = np.zeros((len(userIDS), 27968), dtype=int)
for x in range(len(datasetCOO.data)):
    if datasetCOO.data[x] == 0:
        newURM[datasetCOO.row[x]][datasetCOO.col[x]] = int(1)

newURM

/var/folders/tt/9v6608x550n51759gpcrffsc0000gn/T/ipykernel_15183/2121708647.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  dataset = pd.read_csv('../Input/interactions_and_impressions.csv')


array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [32]:
types = pd.read_csv('../Input/data_ICM_type.csv')
types = types.drop(columns=['data'], axis=1)
types = types.rename(columns={'feature_id': 'type'})
types

item_id  type
0            0     1
1            1     3
2            2     4
3            3     1
4            4     3
...        ...   ...
23086    27963     1
23087    27964     2
23088    27965     1
23089    27966     1
23090    27967     4

[23091 rows x 2 columns]

In [42]:
typesArray = types['type'].to_numpy()
itemsID = types['item_id'].to_numpy()

In [40]:
ICM = np.zeros((27968, 2), dtype=int)

In [43]:
for x in range(len(itemsID)):
    ICM[itemsID[x]][0] = typesArray[x]

In [44]:
ICM

array([[1, 0],
       [3, 0],
       [4, 0],
       ...,
       [1, 0],
       [1, 0],
       [4, 0]])

In [50]:
lenght = pd.read_csv('../Input/data_ICM_length.csv')
lenght = lenght.drop(columns=['feature_id'], axis=1)
lenght = lenght.rename(columns={'data': 'numberOfEpisodes'})
lenght

item_id  numberOfEpisodes
0            0                 1
1            1                 1
2            2                21
3            3                 1
4            4                 1
...        ...               ...
23086    27963                 1
23087    27964                 1
23088    27965                 1
23089    27966                 1
23090    27967                 2

[23091 rows x 2 columns]

In [47]:
lenghtArray = lenght['numberOfEpisodes'].to_numpy()

In [48]:
for x in range(len(itemsID)):
    ICM[itemsID[x]][1] = lenghtArray[x]

In [49]:
ICM

array([[ 1,  1],
       [ 3,  1],
       [ 4, 21],
       ...,
       [ 1,  1],
       [ 1,  1],
       [ 4,  2]])